contracts/median.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.21;
+
+//  ==========  External imports    ==========
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import "hardhat/console.sol";
+
+contract Median is Initializable, OwnableUpgradeable {
+    uint256 public minimumQuorum;
+    bytes32 public currencyPair;
+    uint256 internal lastTimestamp;
+    uint256 internal lastPrice;
+
+    mapping(address => bool) public authorizedNodes;
+    uint256 public authorizedNodesCount;
+
+    mapping(address => bool) public authorizedSigners;
+    uint256 public authorizedSignersCount;
+
+    struct PriceData {
+        uint256 timestamp;
+        uint256 price;
+    }
+
+    PriceData[] public priceHistory;
+
+    // -- EVENTS --
+    event AuthorizedNode(address indexed nodeAddress);
+    event DeauthorizedNode(address indexed nodeAddress);
+    event AuthorizedSigner(address indexed signerAddress);
+    event DeauthorizedSigner(address indexed signerAddress);
+    event MinimumQuorumUpdated(uint256 indexed minimumQuorum);
+    event PriceUpdated(uint256 indexed timestamp, uint256 indexed price);
+
+    // -- ERRORS --
+    error OnlyAuthorizedNodes();
+    error OnlyAuthorizedSigners();
+    error NotEnoughAuthorizedNodes();
+    error InvalidArrayLength();
+    error InvalidSignature();
+    error InvalidTimestamp();
+
+    function initialize(
+        uint256 _minimumQuorum,
+        bytes32 _currencyPair
+    ) public initializer {
+        __Ownable_init_unchained(msg.sender);
+
+        minimumQuorum = _minimumQuorum;
+        currencyPair = _currencyPair;
+    }
+
+    modifier onlyAuthorizedNode() {
+        if (!authorizedNodes[msg.sender]) {
+            revert OnlyAuthorizedNodes();
+        }
+        _;
+    }
+
+    modifier onlyAuthorizedSigner() {
+        if (!authorizedSigners[msg.sender]) {
+            revert OnlyAuthorizedSigners();
+        }
+        _;
+    }
+
+    modifier hasMinimumQuorum() {
+        if (authorizedNodesCount < minimumQuorum) {
```

Comment:
we shouldn't be using a modifier to check for minimumQuorum.

minimumQuorum should be checked when you get the prices.

so inside the `update` function, we'll have:
```
        if (price.length < minimumQuorum) {
```